### DISCUSSION TITLE
getMacLinkCheckAns / getMacDeviceTimeAns question

### DISCUSSION THREAD

**Discussion by michapr** [jgromes/RadioLib#967] [Q&A]
Hi,

using the LoRaWAN_End_Device_Reference.ino I cannot get valid results for the LinkCheck and DeviceTime values.
Connection itself is ok, I can see the transferred data in TTN console.

code (from examples file):
```
    uint8_t margin = 0;
    uint8_t gwCnt = 0;
    if(node.getMacLinkCheckAns(&margin, &gwCnt)) {
      Serial.print(F("[LoRaWAN] LinkCheck margin:\t"));
      Serial.println(margin);
      Serial.print(F("[LoRaWAN] LinkCheck count:\t"));
      Serial.println(gwCnt);
    }

    uint32_t networkTime = 0;
    uint8_t fracSecond = 0;
    if(node.getMacDeviceTimeAns(&networkTime, &fracSecond, true)) {
      Serial.print(F("[LoRaWAN] DeviceTime Unix:\t"));
      Serial.println(networkTime);
      Serial.print(F("[LoRaWAN] LinkCheck second:\t1/"));
      Serial.println(fracSecond);
    }
```
Result:
```
LoRaWAN] LinkCheck margin:	0
[LoRaWAN] LinkCheck count:	0
[LoRaWAN] DeviceTime Unix:	0
[LoRaWAN] LinkCheck second:	1/0
```

What maybe wrong here?

Thanks!

**EDIT:**
another interesting detail:
If add the request
`node.sendMacCommandReq(RADIOLIB_LORAWAN_MAC_LINK_CHECK);`
I will get an answer for
`node.getMacDeviceTimeAns(&networkTime, &fracSecond, true)`
(with same results as above...)

**Comment by HeadBoffin** [comment]
I think you are plumbing the depths of the beta status. I certainly haven't got to ticking that off yet. @StevenCellist & myself are co-ordinating a test rig so we don't end up shuffling around multiple devices & my RF sandbox for 915MHz testing in the EU.

So if you can bear with us - I can patch this in to my test code over the next 48 hours to give some initial feedback.

**Reply by michapr** [reply]
Didn't want to "plumb the depths of the beta status" :)) Only was using an example code for my tests ;)

But will be happy to test more, because I think about to move from Heltec library...

**Reply by HeadBoffin** [reply]
For clarity, LoRaWAN is in beta status - the core functions have been well exercised but we still have a fair number of items yet to test in the wild & the examples haven't entirely kept up with functionality. So at present YMMV. 

I can reproduce your results and @StevenCellist has located at least one issue in the code. With the creation of a test rig we will be able to publish a wiki page of what we are happy is working and then have the basis to look at the next layer of tests.

**Reply by michapr** [reply]
Yes, I know that is in beta status. **And I like very your work, thank you for this!**

I will be happy to make more tests and ask here about any unclear results, if you agree.

**Reply by HeadBoffin** [reply]
No more tests needed at this point, as soon as there is a fix available we'll certainly take up your offer!
